### ISSUE TITLE
Poor ordering in map_overlap

### ISSUE THREAD

**Issue by jakirkham** [dask/dask#3554]
Credits go to @jcrist for discovering this during a recent sprint. Merely opening the issue so we can track it. Placing it on the Dask issue tracker to start, but we can move as needed.

Currently with `map_overlap` operations, it seems the scheduler is preferring computations such that all intermediaries are held in memory before any of the computations will proceed far enough to release chunks of memory. It may be better to prioritize small chains of computations so that memory is released. ATM it is unclear how easy this would be to accomplish or what sort of tradeoffs this will entail. IOW further investigation is required.

Below is a trivial MRE that demonstrates the issue with the corresponding task graph generated by this operation.

<details>
<summary>Code:</summary>

```python
In [1]: import numpy as np

In [2]: import dask.array as da

In [3]: a = da.random.random((12,), chunks=(4,))

In [4]: b = a.map_overlap(lambda e: 2 * e, depth=1)

In [5]: b.visualize(color="order")
Out[5]: <IPython.core.display.Image object>
```

</details>

<details>
<summary>Dask Graph:</summary>

![mydask](https://user-images.githubusercontent.com/3019665/40882706-f1a5f010-669e-11e8-85a7-790e53e258f4.png)

</details>

<details>
<summary>Environment:</summary>

```yaml
name: test
channels:
  - conda-forge
  - defaults
dependencies:
  - appnope=0.1.0=py36_0
  - backcall=0.1.0=py_0
  - blas=1.1=openblas
  - bokeh=0.12.16=py36_0
  - ca-certificates=2018.4.16=0
  - certifi=2018.4.16=py36_0
  - click=6.7=py_1
  - cloudpickle=0.5.3=py_0
  - cycler=0.10.0=py36_0
  - cytoolz=0.9.0.1=py36_0
  - dask=0.17.5=py_0
  - dask-core=0.17.5=py_0
  - decorator=4.3.0=py_0
  - distributed=1.21.8=py36_0
  - freetype=2.8.1=0
  - graphviz=2.38.0=7
  - heapdict=1.0.0=py36_0
  - ipython=6.4.0=py36_0
  - ipython_genutils=0.2.0=py36_0
  - jedi=0.12.0=py36_0
  - jinja2=2.10=py36_0
  - jpeg=9b=2
  - kiwisolver=1.0.1=py36_1
  - libgfortran=3.0.0=0
  - libpng=1.6.34=0
  - libtiff=4.0.9=0
  - locket=0.2.0=py36_1
  - markupsafe=1.0=py36_0
  - matplotlib=2.2.2=py36_1
  - msgpack-python=0.5.6=py36h2d50403_2
  - ncurses=5.9=10
  - numpy=1.14.3=py36_blas_openblas_200
  - openblas=0.2.20=8
  - openssl=1.0.2o=0
  - packaging=17.1=py_0
  - pandas=0.23.0=py36_1
  - parso=0.2.1=py_0
  - partd=0.3.8=py36_0
  - pexpect=4.6.0=py36_0
  - pickleshare=0.7.4=py36_0
  - prompt_toolkit=1.0.15=py36_0
  - psutil=5.4.5=py36_0
  - ptyprocess=0.5.2=py36_0
  - pygments=2.2.0=py36_0
  - pyparsing=2.2.0=py36_0
  - python=3.6.5=1
  - python-dateutil=2.7.3=py_0
  - python-graphviz=0.8.3=py36_0
  - pytz=2018.4=py_0
  - pyyaml=3.12=py36_1
  - readline=7.0=0
  - setuptools=39.2.0=py36_0
  - simplegeneric=0.8.1=py36_0
  - six=1.11.0=py36_1
  - sortedcontainers=2.0.2=py36_0
  - sqlite=3.20.1=2
  - tblib=1.3.2=py36_0
  - tk=8.6.7=0
  - toolz=0.9.0=py_0
  - tornado=5.0.2=py36_0
  - traitlets=4.3.2=py36_0
  - wcwidth=0.1.7=py36_0
  - xz=5.2.3=0
  - yaml=0.1.7=0
  - zict=0.1.3=py_0
  - zlib=1.2.11=h470a237_3
```

</details>

<br>

cc @mrocklin

**Comment by mrocklin** [comment]
Yes, that's a great example.  I found myself wanting to get a picture for what was happening at the larger sale, so I did this:

```python
import numpy as np
import dask.array as da
a = da.random.random(100, chunks=(5,))
b = a.map_overlap(lambda e: 2 * e, depth=1)
b.visualize(color="order", node_attr={'penwidth': '6'})
b.visualize('dask.png', color="order", node_attr={'penwidth': '6'})
```

![dask](https://user-images.githubusercontent.com/306380/40883138-538e1dfa-66c4-11e8-9a82-dc4ae3183a01.png)

But looking into I suspect that you went here too and then scaled down.  The graph you provide seems to be the minimal case.  It looks like we start at `0`, move up until we need to start some other chunk.  We unfortunately choose to proceed with the task on the far right (7) rather than the task in the center (14).  The next step is probably to dive into exactly what motivated that choice.  I suspect that it is preferred because it has more total dependents (7 is a bit deeper than 14) but I'm not sure.  I find it hard to keep the ordering rules in my head long-term.

**Comment by jakirkham** [comment]
It's true that I started bigger and went to the smaller case, but agree it is useful to see the larger case to get a feeling for how this problem behaves at a larger scale.

Do we document the heuristics that inform task ordering somewhere? If not, expect this would be useful to do both for understanding cases like this one and for identifying when the code deviates from our intentions.

**Comment by convexset** [comment]
Hi, was just introduced to this issue, and I'll jump in naively.

I assume the tasks are organized as a DAG whereby each task has:
 - a list of predecessors (inputs)
 - a list of dependents waiting on it (for "notification")
So the scheduler can release anything without outstanding predecessors

The logic involved for the (0), (7), (14) ordering above might be due to a heuristic based on what "completing a released task may allow to be released". So, (0) and (14) are more likely to enable more things to be done before having to wait.

The same applies to the (0), (7), (9).

**Comment by jakirkham** [comment]
If this is interesting to you, @convexset, would encourage you to take a look at issue ( https://github.com/dask/dask-ml/issues/206 ), which likely is struggling with the same problem in a different context (though I could be mistaken). Also PR ( https://github.com/dask/dask/pull/3588 ) includes some examples/tests cases and the beginnings of a discussion about how we might approach them. This would certainly be a fun problem to play with.

**Comment by mrocklin** [comment]
I can solve the minimal problem in the opening comment in #3652 , however this doesn't resolve the underlying issue.

In [this chat](https://gitter.im/sci-sprints/Lobby?at=5b10442ce26c847ac8a85ce0) @emmanuelle provided the following example (slightly modified to remove scikit-learn's apply-parallel):

```python
from skimage import filters
import dask.array as da

ar = da.random.random((10000, 10000), chunks=(512, 512))

res = ar.map_overlap(filters.gaussian, depth=10)

from dask.distributed import Client
client = Client(processes=False)

res.compute()
```

Watching the dashboard it's pretty clear that things aren't being run in a pleasant order.  

An alteration of @jakirkham 's example to 2d also shows a bad result with the proposed PR

```python
import numpy as np
import dask.array as da
a = da.random.random((100, 100), chunks=(20, 20))
b = a.map_overlap(lambda e: 2 * e, depth=1)
b.visualize('dask.pdf', color="order", node_attr={'penwidth': '6'})
```

**Comment by jakirkham** [comment]
Interesting, thanks for the update on this.

Am trying to wrap my head around what these plots are showing. Admittedly, as noted, displaying this stuff in 2D is hard. :) Maybe having a colorbar would help? Perhaps also a perceptually linear colormap would be good (Viridis?). Will try to play with this a bit as well.

Also do we have any idea what is happening at the upper corners in the last image?

**Comment by mrocklin** [comment]
The plots show the order in which a single-threaded scheduler will visit each block of the input and output image.  Seeing that things are smooth is encouraging, it means that we will likely be able to release data as we progress.

I don't think that changing colormaps here would improve things.  This seems to be more a question of understanding what the values mean, rather than a challenge to map between values and colors.

**Comment by mrocklin** [comment]
Any non-smoothness in these images is bad, similar to how non-smoothness is bad for `dask.visualize(..., color='order')`.  Things are more smooth after this PR, which is a good sign.  I don't know why the borders are odd, my guess is that they were de-prioritized by the ordering heuristics because they had fewer dependencies.

**Comment by jakirkham** [comment]
Going to close as this was fixed in `0.18.2`.
